### PR TITLE
[ml-service] Creating training offloading handle

### PR DESCRIPTION
**Pull request by songgot**
- Add ML_SERVICE_TYPE_TRAINING_OFFLOADING to service type

### CODE REVIEW

File: tests/test_models/config/training_offloading_receiver.conf [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+            "description" : "temp description for remote model registeration test using URI"
+        },
+        "pipeline_registration_raw" :
+        {
+            "service-type" : "pipeline_raw",
+            "service-key" : "pipeline_registration_test_key",
+            "description" : "temp description for remote pipeline registeration test using raw file transmission"
+        }
+    }
+}
```
Comment (jaeyun-jung) [inline tests/test_models/config/training_offloading_receiver.conf:39]: plz check eof char

Comment (songgot) [reply tests/test_models/config/training_offloading_receiver.conf:39]: Thank you.

File: c/src/ml-api-service.c [hunk 2]
```diff
@@ -297,8 +300,22 @@ _ml_service_offloading_conf_to_opt (JsonObject * object, const gchar * name,
   for (iter = list; iter != NULL; iter = g_list_next (iter)) {
     const gchar *key = iter->data;
     const gchar *val = json_object_get_string_member (offloading_object, key);
-
     status = ml_option_set (option, key, g_strdup (val), g_free);
+
+    if (g_ascii_strcasecmp (iter->data, "training") == 0) {
```
Comment (gichan-jang) [inline c/src/ml-api-service.c:305]: Let's use `key` instead of `iter->data`. L305, L317, L318

File: c/src/ml-api-service.c [hunk 3]
```diff
@@ -297,8 +300,22 @@ _ml_service_offloading_conf_to_opt (JsonObject * object, const gchar * name,
...
+          offloading_object);
+      if (status != ML_ERROR_NONE) {
+        _ml_error_report_return (status,
+            "Failed to parse the configuration file.");
+      }
+      offloading_s->offloading_type = ML_OFFLOADING_TYPE_TRAINING;
+      continue;
+    }
+    val = json_object_get_string_member (offloading_object, iter->data);
+    status = ml_option_set (option, iter->data, g_strdup (val), g_free);
```
Comment (gichan-jang) [inline c/src/ml-api-service.c:318]: Duplicated lines L302~L303. Pleas Remove these.

File: c/src/ml-api-service.c [hunk 4]
```diff
@@ -355,13 +372,19 @@ _ml_service_offloading_create_json (ml_service_s * mls, JsonObject * object)
     _ml_error_report_return (status, "Failed to create ml-option.");
   }
 
-  status = _ml_service_offloading_conf_to_opt (object, "offloading", option);
+  status = ml_service_offloading_create (mls);
+  if (status != ML_ERROR_NONE) {
+    _ml_error_report ("Failed to create ml-service-offloading.");
```
Comment (gichan-jang) [inline c/src/ml-api-service.c:377]: return here?

File: c/src/ml-api-service.c [hunk 5]
```diff
@@ -355,13 +372,19 @@ _ml_service_offloading_create_json (ml_service_s * mls, JsonObject * object)
...
+      _ml_service_offloading_conf_to_opt (mls, object, "offloading", option);
   if (status != ML_ERROR_NONE) {
     _ml_error_report ("Failed to set ml-option from config file.");
     goto done;
   }
 
-  status = ml_service_offloading_create (mls, option);
+  status = ml_service_offloading_start (mls, option);
   if (status != ML_ERROR_NONE) {
     _ml_error_report ("Failed to create ml-service-offloading.");
```
Comment (gichan-jang) [inline c/src/ml-api-service.c:389]: Let's update the error message

File: tests/capi/unittest_capi_service_offloading.cc [hunk 6]
```diff
@@ -290,12 +290,7 @@ TEST_F (MLOffloadingService, createInvalidParam_n)
   status = ml_option_create (&option_h);
   EXPECT_EQ (ML_ERROR_NONE, status);
 
-  service_h = _ml_service_create_internal (ML_SERVICE_TYPE_OFFLOADING);
-
-  status = ml_service_offloading_create (NULL, option_h);
-  EXPECT_EQ (ML_ERROR_INVALID_PARAMETER, status);
-
-  status = ml_service_offloading_create (service_h, NULL);
+  status = ml_service_offloading_create (service_h);
```
Comment (gichan-jang) [inline tests/capi/unittest_capi_service_offloading.cc:293]: What is the invalid param?

File: c/src/ml-api-service-offloading.h [hunk 7]
```diff
@@ -14,17 +14,61 @@
...
+/**
+ * @brief Enumeration for ml-offloading type.
+ */
+typedef enum
+{
+  ML_OFFLOADING_TYPE_UNKNOWN = 0,
+  ML_OFFLOADING_TYPE_TRAINING,
+
+  ML_OFFLOADING_TYPE_MAX
+} ml_offloading_type_e;
```
Comment (gichan-jang) [inline c/src/ml-api-service-offloading.h:35]: The difference between `ml_offloading_type_e` and `ml_service_offloading_type_e`  is confusing.
Is it difficult to integrate these two struct?
If it is need to be separated, let's consider changing the name.
```
/**
 * @brief Enumeration for ml-offloading service type.
 */
typedef enum
{
  ML_SERVICE_OFFLOADING_TYPE_UNKNOWN = 0,
  ML_SERVICE_OFFLOADING_TYPE_MODEL_RAW,
  ML_SERVICE_OFFLOADING_TYPE_MODEL_URI,
  ML_SERVICE_OFFLOADING_TYPE_PIPELINE_RAW,
  ML_SERVICE_OFFLOADING_TYPE_PIPELINE_URI,
  ML_SERVICE_OFFLOADING_TYPE_REPLY,

  ML_SERVICE_OFFLOADING_TYPE_MAX
} ml_service_offloading_type_e;
```

File: c/src/ml-api-service.c (hunk 2 above)
Comment (songgot) [reply c/src/ml-api-service.c:305]: okay, I will apply it to the last pr.

File: c/src/ml-api-service-offloading.h (hunk 7 above)
Comment (songgot) [reply c/src/ml-api-service-offloading.h:35]: I will consider it, Could you see the last pr?

File: c/src/meson.build [hunk 8]
```diff
@@ -3,7 +3,7 @@ nns_capi_single_srcs = files('ml-api-inference-single.c')
 nns_capi_pipeline_srcs = files('ml-api-inference-pipeline.c')
 nns_capi_service_srcs = files('ml-api-service.c', 'ml-api-service-extension.c', 'ml-api-service-agent-client.c', 'ml-api-service-query-client.c')
 if support_service_offloading
-  nns_capi_service_srcs += files('ml-api-service-offloading.c')
+  nns_capi_service_srcs += files('ml-api-service-offloading.c', 'ml-api-service-training-offloading.c')
```
Comment (gichan-jang) [inline c/src/meson.build:6]: Need to check nntrainer dependency?

File: c/src/ml-api-service.c (hunk 3 above)
Comment (songgot) [reply c/src/ml-api-service.c:318]: It is changed it last pr. Thank you.

File: c/src/ml-api-service.c (hunk 4 above)
Comment (songgot) [reply c/src/ml-api-service.c:377]: Bug fixed in last pr, Thank you.

File: c/src/ml-api-service.c (hunk 5 above)
Comment (songgot) [reply c/src/ml-api-service.c:389]: Sorry, code is changed in last pr (https://github.com/nnstreamer/api/pull/480/commits/421be49e2f24f0089e665a6850f97f0f3ebe6bbc)

**Comment by songgot** [comment]
sqaush to https://github.com/nnstreamer/api/pull/480

**Pull request closed without merging**